bitcoin/src/parse.rs

```diff
@@ -120,3 +120,40 @@ macro_rules! impl_parse_str_through_int {
     }
 }
 pub(crate) use impl_parse_str_through_int;
+
+
+/// Hex Parsing error
+#[derive(Debug, Eq, PartialEq, Clone)]
+#[non_exhaustive]
+pub enum FromHexError<E> {
+    /// The input was not a valid hex string
+    ParseHex(E),
+}
+
+/// Trait that allows types to be initialized from hex strings
+pub trait FromHexStr {
+    type Error;
+
+    /// Parses provided string as hex requiring 0x prefix.
+    ///
+    /// This is intended for user-supplied inputs or already-existing protocols in which 0x prefix is used.
+    fn from_hex_str<S: AsRef<str> + Into<String>>(s: S) -> Result<Self, FromHexError<Self::Error>>
+        where Self: Sized
+    {
+        let val = if !s.as_ref().starts_with("0x") {
+            s.as_ref()
+        } else {
+            s.as_ref().trim_start_matches("0x")
+        };
```

Comment:
This is not done according to the design we discussed in the associated issue. If you ignored it on purpose, please provide your reasoning. Also the doc comment doesn't match.

Reply:
Yea this was kind of an epiphany when changes things up based on your suggestion. The reasoning is the conversion from a hex string shouldn't really care about the prefix, but the implementer should care about how the conversion is performed. This gets rid of one of the error variants so I thought it made sense

Reply:
I'll update the doc comment in the following set of changes

Reply:
> conversion from a hex string shouldn't really care about the prefix

In the associated issue we considered it footgun to make the prefix optional because e.g `10` is not the same thing in dec and hex but is a valid number for both. In this library we're focused on providing APIs that lower the risk of mistakes.

> the implementer should care about how the conversion is performed

For things that are always hex we already have `FromStr`. `FromHexStr` is for cases where it needs to do something else than `FromStr`.

Reply:
I agree with what you're saying but in the issue, providing no 0x prefix creates an error. This error was never constructed in any of the implementations I provided so the linter was going crazy. Then I thought, if we are already allowing the user to provide the implementation for NO prefix, we can just trim it or pass it to that function. This limits the mistakes on the users end but in theory means that someone implementing this or using this functionality may mix-up prefix/no-prefix. I went with the former as it was the happiest path with less errors and got rid of an entire variant

Reply:
Basically, the way its implemented now, any hex string whether prefix or not works with the least amount of errors (which I think is good). And the no_prefix path is there as a way to be implementation specific with no prefix. That way, its very clear what the inputs are to the function. Any hex string, and a hex string without a prefix

Reply:
Thanks for that link @Kixunil! I was just thinking about this the other day but couldn't remember what the principle was called.

And +1 to the general idea that in this library, we definitely prioritize footgun elimination over "it just works TM".